### PR TITLE
feat(state-sync): save state sync header to external storage

### LINKED ISSUE #10459: [State Sync] Upload headers to GCP to be able to work with stateless validation
More info here 
https://docs.google.com/document/d/1ERF6nNrCNeqkMQPeWOqYm7o5eG28HLmDBZVMuSYXe8I/edit

TLDR
State sync cannot run if all nodes only track a subset of shards. A node cannot generate the state sync header containing all incoming receipts for a certain shard.

Proposed hot fix: upload headers to GCP

### PR DESCRIPTION
**Pull request by VanBarbascu**
Upload state sync headers to external storage to enable single shard tracking. 

Fixes #10459

Tested in `testnet` and [Headers were uploaded to GCP](https://console.cloud.google.com/storage/browser/state-parts/chain_id%3Dtestnet/epoch_height%3D2608/epoch_id%3DjdwKefy7yQQW4FiATDcBkNzWmopccsLhFcZ2kj2nJ4y/headers;tab=objects?project=pagoda-public-services&prefix=&forceOnObjectsSortingFiltering=false&pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22)))

### CODE REVIEW

File: chain/client/src/sync/external.rs [hunk 1]
```diff
@@ -7,6 +7,20 @@ use std::sync::Arc;
 use std::time::Duration;
 use std::time::Instant;
 
+pub enum ObjectType {
+    StatePart,
+    StateHeader,
+}
```
Comment (Longarithm) [inline chain/client/src/sync/external.rs:13]: Looking at usecases, I would consider giving StatePart a parameter `StatePart(part_id)`. This would help to deduplicate logic of `external_*_storage_location` by passing `ObjectType` to them and determine path based on that.

Also word "object" looks too abstract when we have context of State. What about `StateFileType` with types `Part` and `Header`?

File: chain/client/src/sync/external.rs [hunk 2]
```diff
@@ -78,35 +92,36 @@ impl ExternalConnection {
...
-    pub async fn put_state_part(
+    /// Uploads the given state part or header to external storage.
+    /// Wrapper for adding is_ok to the metric labels.
+    pub async fn put_obj(
         &self,
-        state_part: &[u8],
+        object: &[u8],
         shard_id: ShardId,
         location: &str,
+        obj_type: ObjectType,
```
Comment (Longarithm) [inline chain/client/src/sync/external.rs:102]: I would make type the first argument to improve readability of calls.

File: nearcore/src/state_sync.rs [hunk 3]
```diff
@@ -264,7 +314,74 @@ async fn state_sync_dump(
...
+                                &external,
+                            )
+                            .await
+                            {
+                                Err(err) => {
+                                    tracing::error!(target: "state_sync_dump", ?err, ?shard_id, "Failed to determine header presence in external storage.");
+                                    false
+                                }
+                                // Header is already stored
+                                Ok(false) => true,
```
Comment (Longarithm) [inline nearcore/src/state_sync.rs:333]: super nit: change `is_state_sync_header_missing_for_epoch` to `is_state_sync_header_stored_for_epoch` to make it `Ok(true) => true` which is more consistent

**Review by Longarithm — CHANGES_REQUESTED** [review]
The logic makes sense, but I have some readability suggestions.

File: chain/client/src/sync/external.rs [hunk 4]
```diff
@@ -115,16 +130,16 @@ impl ExternalConnection {
                     std::fs::create_dir_all(parent_dir)?;
                 }
                 let mut file = std::fs::OpenOptions::new().write(true).create(true).open(&path)?;
-                file.write_all(state_part)?;
-                tracing::debug!(target: "state_sync_dump", shard_id, part_length = state_part.len(), ?location, "Wrote a state part to a file");
+                file.write_all(object)?;
+                tracing::debug!(target: "state_sync_dump", shard_id, part_length = object.len(), ?location, "Wrote a state part to a file");
```
Comment (wacban) [inline chain/client/src/sync/external.rs:134]: Can you also add the type here?

File: chain/client/src/sync/external.rs [hunk 5]
```diff
@@ -78,35 +92,36 @@ impl ExternalConnection {
...
         res
     }
 
-    // Actual implementation.
-    async fn put_state_part_impl(
+    /// Actual implementation.
+    async fn put_obj_impl(
         &self,
-        state_part: &[u8],
+        object: &[u8],
```
Comment (wacban) [inline chain/client/src/sync/external.rs:117]: nit: data

File: chain/client/src/sync/external.rs [hunk 6]
```diff
@@ -201,8 +216,22 @@ impl ExternalConnection {
     }
 }
 
-/// Construct a location on the external storage.
-pub fn external_storage_location(
+/// Construct the header location on the external storage.
+pub fn external_header_storage_location(
```
Comment (wacban) [inline chain/client/src/sync/external.rs:220]: nit: you can refactor to only have one function e.g. `external_file_storage_location` that takes the type as param. Same for the header_filename and part_filename assuming you apply Longarithm's suggestion.

File: chain/client/src/sync/external.rs [hunk 7]
```diff
@@ -222,20 +251,32 @@ pub fn external_storage_location_directory(
...
-        "chain_id={}/epoch_height={}/epoch_id={}/shard_id={}",
-        chain_id, epoch_height, epoch_id.0, shard_id
-    )
+    match obj_type {
+        ObjectType::StatePart => format!(
+            "chain_id={}/epoch_height={}/epoch_id={}/shard_id={}",
+            chain_id, epoch_height, epoch_id.0, shard_id
+        ),
+        ObjectType::StateHeader => format!(
+            "chain_id={}/epoch_height={}/epoch_id={}/headers/shard_id={}",
```
Comment (wacban) [inline chain/client/src/sync/external.rs:272]: I think it's ok for now but generally speaking such migrations should be made backwards compatible. During release the nodes in the network will be running two different versions of neard and so changes in path will likely breaks for the nodes that are on different version than the dumping node.

File: nearcore/src/state_sync.rs [hunk 8]
```diff
@@ -139,6 +143,47 @@ impl StateSyncDumpHandle {
...
+    chain: &Chain,
+) -> anyhow::Result<Vec<u8>> {
+    let header = chain.get_state_response_header(shard_id, sync_hash)?;
+    let mut buffer: Vec<u8> = Vec::new();
+    header.serialize(&mut buffer)?;
+    Ok(buffer)
+}
+
+/// Check if the state sync header exists in the external storage.
+async fn is_state_sync_header_missing_for_epoch(
```
Comment (wacban) [inline nearcore/src/state_sync.rs:159]: nit: It's typically easier to work with positive statements such as "exists" than negative such as "missing". Also it would be more compatible with the comment.

File: nearcore/src/state_sync.rs [hunk 9]
```diff
@@ -264,7 +314,74 @@ async fn state_sync_dump(
                         None
                     }
                     Ok((state_root, num_parts, sync_prev_prev_hash)) => {
-                        match get_missing_part_ids_for_epoch(
+                        // Upload header
```
Comment (wacban) [inline nearcore/src/state_sync.rs:317]: This is really nested and hard to parse, can you refactor by returning early where possible and/or moving to helper methods?

**Review by wacban — COMMENTED** [review]
overall looks good, some readability comments

File: nearcore/src/state_sync.rs (hunk 9 above)
Comment (VanBarbascu) [reply nearcore/src/state_sync.rs:317]: I will try my best to make it as clean as possible but there is a limitation on passing `chain` instance between async functions. 
[This](https://github.com/near/nearcore/commit/c42391481b5e82f0d8b780134e560f1f93978a8b) does not work because of `chain`.

**Review by Longarithm — APPROVED** [review]
